circuit/types/integers/src/add_checked.rs

```diff
@@ -73,37 +73,38 @@ impl<E: Environment, I: IntegerType> AddChecked<Self> for Integer<E, I> {
                 Some(value) => Integer::constant(console::Integer::new(value)),
                 None => E::halt("Integer overflow on addition of two constants"),
             }
-        } else {
-            // TODO: Optimize
+        } else if I::is_signed() {
             // Instead of adding the bits of `self` and `other` directly, the integers are
             // converted into a field elements, and summed, before converting back to integers.
             // Note: This is safe as the field is larger than the maximum integer type supported.
             let sum = self.to_field() + other.to_field();
 
-            // Extract the integer bits from the field element, with a carry bit.
-            let (sum, carry) = match sum.to_lower_bits_le(I::BITS as usize + 1).split_last() {
-                Some((carry, bits_le)) => (Integer::from_bits_le(bits_le), carry.clone()),
+            // Extract the integer bits from the field element, ignoring the carry bit as it is not relevant for signed addition.
+            let sum = match sum.to_lower_bits_le(I::BITS as usize + 1).split_last() {
```

Comment:
```suggestion
            let sum = match sum.to_lower_bits_le(I::BITS as usize) {
```
I think that the call of `to_lower_bits_le` is still allocating an R1CS boolean variable for the carry, even though the pattern below `Some((_, ...))` ignores it at the Rust level. But if instead we pass `I::BITS` to `to_lower_bits_le` and adjust the pattern below, it should only allocate `I::BITS` for the result, and ensure that their weighted sum equals the field value of the sum.

Reply:
This doesn't work because the addition uses the `I::BITS + 1` lower bits of the field element.
`to_lower_bits_le` requires that the you extract all of the nonzero bits from the field.
The interesting thing is that for signed integer addition, overflow is not determined by the carry bit, rather by the signs.
https://www.doc.ic.ac.uk/~eedwards/compsys/arithmetic/index.html

Reply:
You are right, I forgot about the signed case somehow (I was just thinking unsigned).